openupgrade_scripts/scripts/website/16.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,100 @@
+from itertools import chain
+
+from openupgradelib import openupgrade
+from openupgradelib.openupgrade_160 import convert_string_bootstrap_4to5
+
+_xmlids_renames = [
+    (
+        "website.group_website_publisher",
+        "website.group_website_restricted_editor",
+    ),
+    (
+        "website_sale.menu_reporting",
+        "website.menu_reporting",
+    ),
+]
+
+# delete xml xpath for odoo add it again
+_xmlids_delete = [
+    "website.website_configurator",
+    "website.website_menu",
+]
+
+
+def delete_constraint_website_visitor_partner_uniq(env):
+    openupgrade.delete_sql_constraint_safely(
+        env,
+        "website",
+        "website_visitor",
+        "partner_uniq",
+    )
+
+
+def _fill_partner_id_if_null(env):
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE website_visitor
+        SET partner_id = CASE
+            WHEN length(access_token) != 32 THEN CAST(access_token AS integer)
+            ELSE partner_id
+                        END
+        WHERE partner_id IS NULL;
```

Comment:
@duyquyen96 nếu access_token đó tương ứng với id của partner thì mới đúng
```suggestion
        UPDATE website_visitor v
            SET partner_id = p.id
        JOIN res_partner p ON p.id = CAST(v.access_token AS integer)
        WHERE v.partner_id IS NULL AND length(v.access_token) != 32;
```